diag_manager/fms_send_data_statfun.F90

```diff
@@ -0,0 +1,1139 @@
+!! <TT>fms_send_data_stattfun_mod</TT>  defines functions used by the
+!! diag_managers send_data routines tto prepare the filed/data for output.
+!!  The functions include agregation, min and max, etc.
+!!
+!> @file
+!> @brief File for @ref fms_send_data_statfun_mod
+!> @addtogroup fms_diag_dlinked_list_mod
+!> @{
+MODULE fms_send_data_statfun_mod
+
+  use platform_mod
+   USE mpp_mod, ONLY: mpp_pe, mpp_root_pe
+
+   USE fms_mod, ONLY: error_mesg, FATAL, WARNING, NOTE, stdout, stdlog, write_version_number,&
+   & fms_error_handler
+   USE diag_data_mod, ONLY:  input_fields, output_fields, debug_diag_manager
+   use diag_util_mod, ONLY: check_out_of_bounds, update_bounds
+   use fms_diag_weight_procs_mod !!, ONLY :: FmsWeightProcCfg_t
+
+   IMPLICIT NONE
+
+   TYPE STATFUN_IDX_CFG_T
+      INTEGER :: f1,f2,f3,f4
+      INTEGER ::  is, js, ks, ie, je, ke
+      INTEGER :: hi !< halo size in x direction
+      INTEGER :: hj !< halo size in y direction
+   END TYPE STATFUN_IDX_CFG_T
+
+   CONTAINS
+
+
+   FUNCTION AVERAGE_THE_FIELD (diag_field_id, field, out_num, &
+      & mask, weight1, sample, missvalue, missvalue_present, &
+      & l_start, l_end, idx_cfg, err_msg,  err_msg_local ) result( succeded )
+      INTEGER, INTENT(in) :: diag_field_id
+      REAL, DIMENSION(:,:,:), INTENT(in) :: field
+      INTEGER, INTENT(in) :: out_num
+      LOGICAL, DIMENSION(:,:,:), INTENT(in), OPTIONAL :: mask
+      REAL, INTENT(in) :: weight1
+      INTEGER, INTENT(in) :: sample
+      REAL, INTENT(in), OPTIONAL :: missvalue
+      LOGICAL, INTENT(in) :: missvalue_present
+      INTEGER, DIMENSION(3), INTENT(in)  :: l_start !< local start indices on 3 axes for regional output
+      INTEGER, DIMENSION(3), INTENT(in)  :: l_end !< local end indices on 3 axes for regional output
+      CHARACTER(len=*), INTENT(inout), OPTIONAL :: err_msg
+      CHARACTER(len=*), INTENT(inout) :: err_msg_local
+      TYPE(STATFUN_IDX_CFG_T), INTENT(in) :: idx_cfg
+
+      LOGICAL :: succeded
+
+      !!LOGICAL, ALLOCATABLE, DIMENSION(:,:,:) :: oor_mask
+      CHARACTER(len=128):: error_string
+
+      ! Power value for rms or pow(x) calculations
+      INTEGER :: pow_value, ksr, ker, is, js, ks, ie, je, ke, hi, hj, f1, f2, f3, f4
+      LOGICAL :: phys_window , need_compute , reduced_k_range
+      !!TODO: name change PHYS_WINDOWS -->> ? OMP subsetted data, See output_fields
+
+      TYPE(FmsWeightProcCfg_t), allocatable :: weight_procs
+
+      INTEGER :: i, j, k,  i1, j1, k1
+
+      INTEGER :: numthreads
+      INTEGER :: active_omp_level
+#if defined(_OPENMP)
+      INTEGER :: omp_get_num_threads !< OMP function
+      INTEGER :: omp_get_level !< OMP function
+#endif
+
+
+      pow_value = output_fields(out_num)%pow_value
+      ALLOCATE(weight_procs)
+      call weight_procs%initialize (pow_value)
+
+      phys_window = output_fields(out_num)%phys_window
+      need_compute = output_fields(out_num)%need_compute
+      reduced_k_range = output_fields(out_num)%reduced_k_range
```

Comment:
I cringed looking at this.  This is using the old object.  You will have to recreate this functions using the new fms_diag_object.  I mentioned putting a big `IF` in the diag_manager.F90, but it's possible that you could have the `if (use_moodern_diag)` statements in here and add the support for the diag_object.

Reply:
Yes you are right. I was thinking of putting more info (e.g.. phys_window, need_compute, ...) into the idx_cfg.
But the old diag code and the new code would have to set idx_cfg.  With that change, and passing in the fields,
and using class(*), then maybe both old and new code can use a common fms_senddata_statfun.F90.
Not sure about all of the details right now.